### PR TITLE
Add blog article about the new collections performance

### PR DESCRIPTION
**Pull request by julienrf**

### CODE REVIEW

File: blog/_posts/2018-02-07-collections-performance.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+them to be slower since the execution path, when calling an operation, can be made
+exactly the same as in the old collections.
+
+## Conclusion
+
+This article studied the performance of the new collections. I’ve reported that view
+based operation implementations are about 25% slower than builder based implementations
+and I’ve explained how we restored builder based implementations on strict collections.
+
+I expect the new collections to be as fast or slightly faster than the previous collections.
```
Comment (olafurpg) [inline blog/_posts/2018-02-07-collections-performance.md]: s/fast or slightly faster/equally fast or slightly faster/s

?

File: blog/_posts/2018-02-07-collections-performance.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+Indeed, we took advantage of the rewrite to apply some more optimizations here and
+again.
+
+More significant performance improvements can be achieved by using different
+data structures. For instance, we recently
+[merged](https://github.com/scala/collection-strawman/pull/342)
+a completely new implementation of immutable `Set` and `Map` based on [compressed
+hash-array mapped prefix-trees](https://michael.steindorfer.name/publications/oopsla15.pdf).
+This data structure has a smaller memory footprint than the old `HashSet` and `HashMap`,
```
Comment (olafurpg) [inline blog/_posts/2018-02-07-collections-performance.md]: Can you include concrete numbers here? Sounds amazing!

File: blog/_posts/2018-02-07-collections-performance.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+[here](https://gist.github.com/julienrf/f1cb2b062cd9783a35e2f35778959c76).
+
+Since operation implementations end up being the same, why do we get better performance
+at all? Well, these numbers are specific to `Vector`, and are due to the fact that
+we more agressively inlined a few critical methods. I don’t expect the new collections
+to be *always* 20% faster than the old collections. However, there is no reason for
+them to be slower since the execution path, when calling an operation, can be made
+exactly the same as in the old collections.
+
+## Conclusion
```
Comment (olafurpg) [inline blog/_posts/2018-02-07-collections-performance.md]: I would include "correct first" somewhere in the conclusion, it's an interesting point worth repeating.

File: blog/_posts/2018-02-07-collections-performance.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
+---
+layout: blog-detail
+post-type: blog
+by: Julien Richard-Foy
+title: On Performance of the New Collections
+---
+
+In a [previous blog post](/blog/2017/11/28/view-based-collections.html), I explained
+how [Scala 2.13’s new collections](http://www.scala-lang.org/blog/2017/02/28/collections-rework.html)
+have been designed so that the default implementations of transformation operations work
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: is designed such

File: blog/_posts/2018-02-07-collections-performance.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+with both strict and non-strict types of collections. In essence, we abstract over
+the evaluation mode (strict or non strict) of concrete collection types.
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: non-strict

File: blog/_posts/2018-02-07-collections-performance.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+For reference, the source code of the new collections is available in
+[this GitHub repository](https://github.com/scala/collection-strawman).
+
+## Overhead Of View Based Implementations
+
+Let’s be clear: the view based implementations are in general slower than their
+builder based versions. How much slower exactly varies with the type of collection
+(e.g. `List`, `Vector`, `Set`), the operation (e.g. `map`, `flatMap`, `filter`)
+and the number of elements in the collection. In my benchmark on `Vector`, on
+the `map`, `filter` and `flatMap` operations, with 1 element to 7 million of
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: with 1 to 7 million of elements

File: blog/_posts/2018-02-07-collections-performance.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+about possible performance implications of having more levels of abstraction than before.
+
+This blog article gives more information about the overhead of the
+collections’ view-based design and our solution to remove that
+overhead.
+
+For reference, the source code of the new collections is available in
+[this GitHub repository](https://github.com/scala/collection-strawman).
+
+## Overhead Of View Based Implementations
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: View-based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+~~~ scala
+trait Vector[+A] extends IndexedSeq[A]
+  with IndexedSeqOps[A, Vector, Vector[A]]
+  with StrictOptimizedSeqOps[A, Vector, Vector[A]]
+~~~
+
+Here we use `StrictOptimizedSeqOps`, which is a specialization of `StrictOptimizedIterableOps`
+for `Seq` collections.
+
+## Is The View Based Design Worth It?
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: View-Based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+This blog article gives more information about the overhead of the
+collections’ view-based design and our solution to remove that
+overhead.
+
+For reference, the source code of the new collections is available in
+[this GitHub repository](https://github.com/scala/collection-strawman).
+
+## Overhead Of View Based Implementations
+
+Let’s be clear: the view based implementations are in general slower than their
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: view-based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+builder based versions. How much slower exactly varies with the type of collection
+(e.g. `List`, `Vector`, `Set`), the operation (e.g. `map`, `flatMap`, `filter`)
+and the number of elements in the collection. In my benchmark on `Vector`, on
+the `map`, `filter` and `flatMap` operations, with 1 element to 7 million of
+elements, I measured an average slowdown of 25%.
+
+## How To Fix That Performance Regression?
+
+Our solution is simply to go back to builder based implementations for strict collections: we
+override the default view based implementations with more efficient builder based
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: view-based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+![](/resources/img/new-collections-performance-filter.png)
+
+![](/resources/img/new-collections-performance-map.png)
+
+![](/resources/img/new-collections-performance-flatMap.png)
+
+These charts show the execution time (vertically) of the `filter`, `map` and `flatMap`
+operations, according to the number of elements (horizontally). Note that scales are
+logarithmic in both axis. The blue line shows the performance of the old `Vector`,
+the green line shows the performance of the new `Vector` if it used only view based
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: view-based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  with StrictOptimizedSeqOps[A, Vector, Vector[A]]
+~~~
+
+Here we use `StrictOptimizedSeqOps`, which is a specialization of `StrictOptimizedIterableOps`
+for `Seq` collections.
+
+## Is The View Based Design Worth It?
+
+In my previous article I explained that a drawback of the old builder based design was that,
+on non strict collections (e.g. `Stream` or `View`), we had to carefully override all the
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: non-strict

File: blog/_posts/2018-02-07-collections-performance.md [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+default implementations of transformation operations to make them non strict.
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: non-strict

File: blog/_posts/2018-02-07-collections-performance.md [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+Now it seems that the situation is just reversed: the default implementations work well
+with non strict collections, but we have to override them in strict collections.
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: non-strict

File: blog/_posts/2018-02-07-collections-performance.md [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+So, is the new design worth it? To answer this question I will quote a comment posted
+by Stefan Zeiger [here](https://www.reddit.com/r/scala/comments/7g52cy/let_them_be_lazy/dqixt8d/):
+
+> The lazy-by-default approach is mostly beneficial when you're implementing lazy
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: most beneficial

File: blog/_posts/2018-02-07-collections-performance.md [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+![](/resources/img/new-collections-performance-filter.png)
+
+![](/resources/img/new-collections-performance-map.png)
+
+![](/resources/img/new-collections-performance-flatMap.png)
+
+These charts show the execution time (vertically) of the `filter`, `map` and `flatMap`
+operations, according to the number of elements (horizontally). Note that scales are
+logarithmic in both axis. The blue line shows the performance of the old `Vector`,
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: both axes

File: blog/_posts/2018-02-07-collections-performance.md [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+the green line shows the performance of the new `Vector` if it used only view based
+implementations, and the red line shows the actual performance of the new `Vector`
+(with strict optimized implementations). Benchmark source code and numbers can be found
+[here](https://gist.github.com/julienrf/f1cb2b062cd9783a35e2f35778959c76).
+
+Since operation implementations end up being the same, why do we get better performance
+at all? Well, these numbers are specific to `Vector`, and are due to the fact that
+we more agressively inlined a few critical methods. I don’t expect the new collections
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: aggressively

File: blog/_posts/2018-02-07-collections-performance.md [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+to be *always* 20% faster than the old collections. However, there is no reason for
+them to be slower since the execution path, when calling an operation, can be made
+exactly the same as in the old collections.
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: exactly the same => the same

File: blog/_posts/2018-02-07-collections-performance.md [hunk 19]
```diff
@@ -0,0 +1,136 @@
+---
+layout: blog-detail
+post-type: blog
+by: Julien Richard-Foy
+title: On Performance of the New Collections
+---
+
+In a [previous blog post](/blog/2017/11/28/view-based-collections.html), I explained
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: I would group all those paragraphs. until `Overhead Of View Based Implementations`

File: blog/_posts/2018-02-07-collections-performance.md (hunk 9 above)
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: Let’s be clear,

File: blog/_posts/2018-02-07-collections-performance.md [hunk 20]
```diff
@@ -0,0 +1,136 @@
...
+
+For reference, the source code of the new collections is available in
+[this GitHub repository](https://github.com/scala/collection-strawman).
+
+## Overhead Of View Based Implementations
+
+Let’s be clear: the view based implementations are in general slower than their
+builder based versions. How much slower exactly varies with the type of collection
+(e.g. `List`, `Vector`, `Set`), the operation (e.g. `map`, `flatMap`, `filter`)
+and the number of elements in the collection. In my benchmark on `Vector`, on
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: In our

File: blog/_posts/2018-02-07-collections-performance.md [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+the `map`, `filter` and `flatMap` operations, with 1 element to 7 million of
+elements, I measured an average slowdown of 25%.
+
+## How To Fix That Performance Regression?
+
+Our solution is simply to go back to builder based implementations for strict collections: we
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: builder-based

File: blog/_posts/2018-02-07-collections-performance.md (hunk 10 above)
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: builder-based

File: blog/_posts/2018-02-07-collections-performance.md [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+override the default view based implementations with more efficient builder based
+ones. We actually and up with the same implementations as in the old collections.
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: `We actually and up with` ?

File: blog/_posts/2018-02-07-collections-performance.md [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+  with IndexedSeqOps[A, Vector, Vector[A]]
+  with StrictOptimizedSeqOps[A, Vector, Vector[A]]
+~~~
+
+Here we use `StrictOptimizedSeqOps`, which is a specialization of `StrictOptimizedIterableOps`
+for `Seq` collections.
+
+## Is The View Based Design Worth It?
+
+In my previous article I explained that a drawback of the old builder based design was that,
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: In my previous article,

Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: This sentence could be split into two.

In my previous article, I explained that a drawback of the old builder based design. On non-strict collections (e.g. Stream or View), we had to carefully override all the default implementations of transformation operations to make them non-strict.

File: blog/_posts/2018-02-07-collections-performance.md [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+So, is the new design worth it? To answer this question I will quote a comment posted
+by Stefan Zeiger [here](https://www.reddit.com/r/scala/comments/7g52cy/let_them_be_lazy/dqixt8d/):
+
+> The lazy-by-default approach is mostly beneficial when you're implementing lazy
+> collections because you don't have to override pretty much everything or get
+> incorrect semantics. The reverse risk is smaller: If you don't override a lazy
+> implementation for a strict collection type you only suffer a small performance
+> impact but it's still correct.
+
+In short: implementations are **correct first** in the new design but you might want to
```
Comment (MasseGuillaume) [inline blog/_posts/2018-02-07-collections-performance.md]: In short,

**Comment by julienrf** [comment]
Just a heads up that I defined the release date for this article to today. Let me know if you need more time to review or if I should merge it.

File: blog/_posts/2018-02-07-collections-performance.md [hunk 25]
```diff
@@ -0,0 +1,139 @@
...
+
+![](/resources/img/new-collections-performance-flatMap.png)
+
+These charts show the execution time (vertically) of the `filter`, `map` and `flatMap`
+operations, according to the number of elements (horizontally). Note that scales are
+logarithmic in both axes. The blue line shows the performance of the old `Vector`,
+the green line shows the performance of the new `Vector` if it used only view based
+implementations, and the red line shows the actual performance of the new `Vector`
+(with strict optimized implementations). Benchmark source code and numbers can be found
+[here](https://gist.github.com/julienrf/f1cb2b062cd9783a35e2f35778959c76).
```
Comment (lrytz) [inline blog/_posts/2018-02-07-collections-performance.md]: In the logarithmic scale it's not obvious to see the speedup factor. could you say something about it?

Comment (julienrf) [reply blog/_posts/2018-02-07-collections-performance.md]: I mention that the scale is logarithmic and I’ve given the speedup factors in the text (view-based implementations are 25% slower than builder-based ones, and the new collections are 35% faster than the old collections).

What should I add?

Comment (lrytz) [reply blog/_posts/2018-02-07-collections-performance.md]: Ah I missed the mention of 35 above the charts. I guess my concern is that the graphs the thing that really jups at people's eyes. The current for with the logarithmic scale doesn't really visualize the nice improvements. Maybe you could do some bars on a linear scale, normalize the old collection to `1`, and add time values to the bars? Something like https://docs.google.com/spreadsheets/d/1Jcw3iG5sx_Xo-3svjb_qmB4Fprz5sEX7nCFpCd4EcWQ/edit?usp=sharing

Comment (julienrf) [reply blog/_posts/2018-02-07-collections-performance.md]: Yeah I was considering of doing something exactly like that! OK, let’s do it then.

File: blog/_posts/2018-02-07-collections-performance.md [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+by Stefan Zeiger [here](https://www.reddit.com/r/scala/comments/7g52cy/let_them_be_lazy/dqixt8d/):
+
+> The lazy-by-default approach is mostly beneficial when you're implementing lazy
+> collections because you don't have to override pretty much everything or get
+> incorrect semantics. The reverse risk is smaller: If you don't override a lazy
+> implementation for a strict collection type you only suffer a small performance
+> impact but it's still correct.
+
+In short, implementations are **correct first** in the new design but you might want to
+override them for performance reasons on strict collections.
```
Comment (SethTisue) [inline blog/_posts/2018-02-07-collections-performance.md]: This sentence seems like the crux of the whole post. I suggest saying something this right at the top. Many people will read partway in and then bail, so you should hit the takeaways early, _then_ proceed with more detailed explanations.

File: blog/_posts/2018-02-07-collections-performance.md [hunk 27]
```diff
@@ -0,0 +1,139 @@
...
+
+## Performance Comparison With 2.12’s Collections
+
+Talking about performance, how performant are the new collections compared to the old ones?
```
Comment (SethTisue) [inline blog/_posts/2018-02-07-collections-performance.md]: Again, I suggest including some brief answer to this very near the top of the whole blog post.

**Comment by julienrf** [comment]
Alright, thanks all for your reviews, I’ve improved the charts and mentioned the takeaways in the first sections. I can squash the commits but I think the “squash and merge” button should just work.

**Comment by julienrf** [comment]
I updated the publication date to today.

**Comment by lrytz** [comment]
Looks good!

**Pull request merged**